### PR TITLE
more epipolar metrics (+`linalg` utility)

### PR DESCRIPTION
**Pull request by nitaifingerhut**
#### Changes
`linalg`: adding utility for measuring the distance from point to line.
epipolar metrics: adding "one-sided" metrics.

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Review by ducha-aiki — APPROVED** [review]
Looks great! Please, don't forget add to the docs
https://github.com/kornia/kornia/blob/master/docs/source/geometry.epipolar.rst

**Comment by nitaifingerhut** [comment]
@ducha-aiki thanks! updated docs

File: kornia/geometry/epipolar/_metrics.py [hunk 1]
```diff
@@ -111,3 +113,66 @@ def symmetrical_epipolar_distance(
...
+         avoid ambiguity with torch.nn.functional.
+
+    Returns:
+        the computed Symmetrical distance with shape :math:`(*, N)`.
+    """
+    KORNIA_CHECK_IS_TENSOR(pts1)
+    KORNIA_CHECK_IS_TENSOR(pts2)
+    KORNIA_CHECK_IS_TENSOR(Fm)
+
+    if (len(Fm.shape) < 3) or not Fm.shape[-2:] == (3, 3):
```
Comment (edgarriba) [inline kornia/geometry/epipolar/_metrics.py:138]: you might want to implement `KORNIA_CHECK_SHAPE`
https://github.com/kornia/kornia/blob/master/kornia/testing/__init__.py#L186

File: kornia/geometry/linalg.py [hunk 2]
```diff
@@ -200,5 +197,31 @@ def transform_points(trans_01: torch.Tensor, points_1: torch.Tensor) -> torch.Te
     points_0 = points_0.reshape(shape_inp)
     return points_0
 
+
+def points_lines_distances(pts: torch.Tensor, lines: torch.Tensor):
```
Comment (edgarriba) [inline kornia/geometry/linalg.py:201]:
```suggestion
def point_line_distance(point: Tensor, line: Tensor) -> Tensor:
```
I'd use as singular which is at least the minimum supported and same convention as other functions
example here: https://github.com/kornia/kornia/blob/master/kornia/utils/draw.py#L26

File: kornia/geometry/linalg.py [hunk 3]
```diff
@@ -200,5 +197,31 @@ def transform_points(trans_01: torch.Tensor, points_1: torch.Tensor) -> torch.Te
...
+    KORNIA_CHECK_IS_TENSOR(pts)
+    KORNIA_CHECK_IS_TENSOR(lines)
+
+    if not pts.shape[-1] in (2, 3):
+        raise ValueError(f"pts must be a (*, 2 or 3) tensor. Got {pts.shape}")
+
+    if not lines.shape[-1] == 3:
+        raise ValueError(f"lines must be a (*, 3) tensor. Got {lines.shape}")
+
+    numerator = torch.abs(lines[..., 0] * pts[..., 0] + lines[..., 1] * pts[..., 1] + lines[..., 2])
```
Comment (edgarriba) [inline kornia/geometry/linalg.py:220]:
```suggestion
    numerator = (lines[..., 0] * pts[..., 0] + lines[..., 1] * pts[..., 1] + lines[..., 2]).abs()
```
more pythonic

File: kornia/geometry/linalg.py [hunk 4]
```diff
@@ -200,5 +197,31 @@ def transform_points(trans_01: torch.Tensor, points_1: torch.Tensor) -> torch.Te
...
+    denominator = torch.linalg.norm(lines[..., :2], dim=-1)
```
Comment (edgarriba) [inline kornia/geometry/linalg.py:221]:
```suggestion
    denominator = lines[..., :2].norm(dim=-1)
```
i'd like more to remove `dim` kwarg to make it more generic in future with numpy or jax though

File: kornia/geometry/linalg.py [hunk 5]
```diff
@@ -200,5 +197,31 @@ def transform_points(trans_01: torch.Tensor, points_1: torch.Tensor) -> torch.Te
...
+
+    return numerator / denominator
```
Comment (edgarriba) [inline kornia/geometry/linalg.py:223]:
```suggestion
    return numerator / (denominator + eps)
```
to make safe division

File: kornia/geometry/epipolar/_metrics.py [hunk 6]
```diff
@@ -111,3 +113,66 @@ def symmetrical_epipolar_distance(
     if squared:
         return out
     return (out + eps).sqrt()
+
+
+def left_to_right_epipolar_distance(pts1: Tensor, pts2: Tensor, Fm: Tensor) -> Tensor:
+    r"""Return one-sided epipolar distance for correspondences given the fundamental matrix.
```
Comment (edgarriba) [inline kornia/geometry/epipolar/_metrics.py:119]:
```suggestion
    r"""Return one-sided epipolar distance for correspondences given the fundamental matrix.

```

File: kornia/geometry/epipolar/_metrics.py [hunk 7]
```diff
@@ -111,3 +113,66 @@ def symmetrical_epipolar_distance(
...
+    This method measures the distance from points in the right images to the epilines of the
+    corresponding points in the left images as they reflect in the right images.
+
+    Args:
+       pts1: correspondences from the left images with shape
+         (*, N, 2 or 3). If they are not homogeneous, converted automatically.
```
Comment (edgarriba) [inline kornia/geometry/epipolar/_metrics.py:125]:
```suggestion
         :math:`(*, N, 2 or 3)`. If they are not homogeneous, converted automatically.
```

File: kornia/geometry/epipolar/_metrics.py [hunk 8]
```diff
@@ -111,3 +113,66 @@ def symmetrical_epipolar_distance(
...
+        pts1 = convert_points_to_homogeneous(pts1)
+
+    F_t: Tensor = Fm.transpose(dim0=-2, dim1=-1)
+    line1_in_2: Tensor = pts1 @ F_t
+
+    return points_lines_distances(pts2, line1_in_2)
+
+
+def right_to_left_epipolar_distance(pts1: Tensor, pts2: Tensor, Fm: Tensor) -> Tensor:
+    r"""Return one-sided epipolar distance for correspondences given the fundamental matrix.
```
Comment (edgarriba) [inline kornia/geometry/epipolar/_metrics.py:151]:
```suggestion
    r"""Return one-sided epipolar distance for correspondences given the fundamental matrix.

```

**Review by edgarriba — COMMENTED** [review]
one meta-though here: Should we consider to move this under `kornia.metrics` ? not sure what's best in this case

**Comment by nitaifingerhut** [comment]
not sure where it should be as well... but as long as the epipolar metrics are there, the new ones should be too. also epipolar metrics are very very specific so I think it's fine there. it's not like say L1 which is a "general" metric for any inputs.
